server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java

```diff
@@ -0,0 +1,102 @@
+import org.elasticsearch.repositories.RepositoryData;
+
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.common.Strings.collectionToDelimitedStringWithLimit;
+import static org.elasticsearch.health.HealthStatus.GREEN;
+import static org.elasticsearch.health.HealthStatus.RED;
+import static org.elasticsearch.health.ServerHealthComponents.SNAPSHOT;
+
+/**
+ * This indicator reports health for snapshot repositories.
+ *
+ * Indicator will report RED status when any of snapshot repositories is marked as corrupted.
+ * Data might not be backed up in such cases.
+ *
+ * Corrupted repository most likely need to be manually cleaned and a new snapshot needs to be created from scratch.
+ */
+public class RepositoryIntegrityHealthIndicatorService implements HealthIndicatorService {
+
+    public static final String NAME = "repository integrity";
+
+    private final ClusterService clusterService;
+
+    public RepositoryIntegrityHealthIndicatorService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public String component() {
+        return SNAPSHOT;
+    }
+
+    @Override
+    public HealthIndicatorResult calculate() {
+        var snapshotMetadata = clusterService.state().metadata().custom(RepositoriesMetadata.TYPE, RepositoriesMetadata.EMPTY);
+
+        if (snapshotMetadata.repositories().isEmpty()) {
+            return createIndicator(GREEN, "No repositories configured", HealthIndicatorDetails.EMPTY);
+        }
+
+        var corrupted = snapshotMetadata.repositories()
+            .stream()
+            .filter(repository -> repository.generation() == RepositoryData.CORRUPTED_REPO_GEN)
+            .map(RepositoryMetadata::name)
+            .toList();
+
+        var totalRepositories = snapshotMetadata.repositories().size();
+        var corruptedRepositories = corrupted.size();
+
+        if (corrupted.isEmpty()) {
+            return createIndicator(
+                GREEN,
+                "No corrupted repositories",
+                new SimpleHealthIndicatorDetails(Map.of("total-repositories", totalRepositories))
+            );
+        }
+
+        return createIndicator(
+            RED,
+            createCorruptedRepositorySummary(corrupted),
+            new SimpleHealthIndicatorDetails(
+                Map.of(
+                    "total-repositories",
+                    totalRepositories,
+                    "corrupted-repositories",
+                    corruptedRepositories,
+                    "corrupted",
+                    corrupted.stream().limit(10).toList()
+                )
+            )
+        );
+    }
+
+    private static String createCorruptedRepositorySummary(List<String> corrupted) {
+        var message = new StringBuilder().append("Detected [").append(corrupted.size()).append("] corrupted repositories ");
```

Comment:
Consider using 
```
var message = new StringBuilder("""
            Detected [%s] corrupted repositories \
            """.formatted(corrupted.size()));
```
here. I believe the support for `formatted` was allowed in the forbidden API check during the work on text blocks.